packages/shared/src/guards.ts

```diff
@@ -421,34 +421,37 @@ export function isComplexValue(
   return type.class === DTypeClass.Complex;
 }
 
-function assertPrimitiveValue<D extends Dataset>(
-  dataset: D,
+function assertPrimitiveValue<T extends DType>(
+  type: T,
   value: unknown,
 ): asserts value is Primitive<DType> {
-  if (hasNumericType(dataset)) {
+  if (isNumericType(type)) {
     assertNum(value);
-  } else if (hasStringType(dataset)) {
+  } else if (isStringType(type)) {
     assertStr(value);
-  } else if (hasBoolType(dataset)) {
+  } else if (isBoolType(type)) {
     assertNumOrBool(value);
-  } else if (hasComplexType(dataset)) {
+  } else if (isComplexType(type)) {
     assertComplex(value);
+  } else if (isCompoundType(type)) {
+    assertArray(value);
   }
 }
 
 export function assertDatasetValue<D extends Dataset<ScalarShape | ArrayShape>>(
   value: unknown,
   dataset: D,
 ): asserts value is Value<D> {
-  if (hasArrayShape(dataset)) {
+  const { type } = dataset;
+
+  if (hasScalarShape(dataset)) {
+    assertPrimitiveValue(type, value);
+  } else {
```

Comment:
Swapping the logic because `type` was narrowed down to `never` in the `else` branch due to the fact that `ScalarShape` ⊂ `ArrayShape` (`[]` ⊂ `number[]`). I have a note somewhere to try to change `ArrayShape` to `[number, ...number[]]`, though I have to check if it's worth the effort.